### PR TITLE
Network generation

### PR DESCRIPTION
**Pull request by leotrs**
I studied the code line by line, added comments and refactored a few bits. Fixed a couple of bugs. Changed the console output to make it more descriptive.

I submit a PR now because both the code and the inner workings of the simulation behave almost exactly the same as in master, whereas the next changes I'll be making from now on will change how the network is created.

### CODE REVIEW

File: abm_sketch.py [hunk 1]
```diff
@@ -16,17 +16,19 @@
...
+
+y_pos_dist = norm(300, 10)
+
 cluster_x_dists = {
-    'A': uniform(0, 50), 
-    'B': uniform(30, 50),
-    'C': uniform(60, 50)
+    'A': uniform(0, 40),
+    'B': uniform(20, 60),
+    'C': uniform(40, 80)
```
Comment (bhtucker) [inline abm_sketch.py:25]: the second argument is the width -- these are increasingly broad dists?

File: abm_sketch.py [hunk 2]
```diff
@@ -35,182 +37,330 @@
...
+            plt.scatter([p.x for p in class_points], [p.y for p in class_points], c=color, s=size)
+
+        # Draw the connections
+        if self.connectivity_matrix is not None:
+            for start_ix, connections in enumerate(self.connectivity_matrix):
+                for connect_ix, connected in enumerate(connections):
+                    if connected and connect_ix != start_ix:
+                        plt.plot(*zip(
+                            (pts[start_ix].x, pts[start_ix].y),
+                            (pts[connect_ix].x, pts[connect_ix].y)),
```
Comment (bhtucker) [inline abm_sketch.py:143]: Another feature of 'edge plotting' I have in mind is encoding messaging probability as line thickness or something. So we'll likely want an interface to edge plotting before long

File: abm_sketch.py (hunk 1 above)
Comment (leotrs) [inline abm_sketch.py:25]: Messed that up, but fixed it at a later commit.

File: abm_sketch.py [hunk 3]
```diff
@@ -35,182 +37,330 @@
...
+                self.task_attempt_map[task.id].append(sender)
+
+            if sender in self.sent:
+                print('popping %s' % self.sent.pop(self.sent.index(sender)))
+
+        # If I have passed the message to all of my neighbors and I still
+        # received it back, bail out.
         if set(self.adjacencies).issubset(set(self.task_attempt_map[task.id])):
+            # FIXME: this catches cycles eventually, but also catches randomly
+            # followed closed paths
```
Comment (bhtucker) [inline abm_sketch.py:238]: yeah its not really 'cycles' so lets fix the printed message too

File: abm_sketch.py [hunk 4]
```diff
@@ -35,182 +37,330 @@
...
+        """
+
         self.value += value
-        print(self.value, self.index, self.adjacencies)
+        print("")
+        print("awarding %.2f to %d" % (self.value, self.index))
+        print("Map: " + str(dict(self.task_attempt_map)))
+        print("Adj before: " + str(self.adjacencies))
+
+        self.learn()
```
Comment (bhtucker) [inline abm_sketch.py:292]: great

File: abm_sketch.py (hunk 2 above)
Comment (leotrs) [inline abm_sketch.py:143]: For the new network generation mechanism I have working now, the location of the point is irrelevant. So I was looking into modifying the locations of each point every time and draw points closer as their likelihood to share a message increases.

But leaving them in place and using the edges sounds easier.

**Comment by bhtucker** [comment]
This is great, and should prime us to be able to refactor and write tests.

The glaring thing in the project now is the need to clean up population construction in general, so I'll merge and let you continue with that for another PR

**Pull request merged**